### PR TITLE
Completed fourth part of Diploma.

### PR DESCRIPTION
**Pull request by VlasovM**
Added own annotation to validate Email, password and name, when user register.
Added Spring security: view your posts, login, logout.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -54,13 +49,33 @@
...
             <artifactId>lombok</artifactId>
             <version>${lombok.version}</version>
             <scope>provided</scope>
         </dependency>
 
+        <!-- https://mvnrepository.com/artifact/org.hibernate.validator/hibernate-validator -->
+        <dependency>
+            <groupId>org.hibernate.validator</groupId>
+            <artifactId>hibernate-validator</artifactId>
+            <version>6.2.0.Final</version>
```
Comment (Klepus) [inline pom.xml:76]: версия

File: src/main/java/com/javlasov/blog/controller/ApiGeneralController.java [hunk 2]
```diff
@@ -38,13 +38,12 @@ private ResponseEntity<SettingsResponse> settings() {
...
-    private ResponseEntity<TagResponse> tag(@RequestParam(required = false, defaultValue = "") String query) {
+    private ResponseEntity<TagResponse> tag(@RequestParam(required = false) Optional<String> query) {
         return ResponseEntity.ok(tagService.tag(query));
     }
 
     @GetMapping("/calendar")
-    private ResponseEntity<CalendarResponse> calendar(@RequestParam(required = false, defaultValue = "0") int year) {
-        return (year == 0) ? ResponseEntity.ok(calendarService.calendar(LocalDate.now().getYear()))
-                : ResponseEntity.ok(calendarService.calendar(year));
+    private ResponseEntity<CalendarResponse> calendar(@RequestParam(required = false) Optional<Integer> year) {
```
Comment (Klepus) [inline src/main/java/com/javlasov/blog/controller/ApiGeneralController.java:46]: private?

File: src/main/java/com/javlasov/blog/controller/ApiGeneralController.java [hunk 3]
```diff
@@ -38,13 +38,12 @@ private ResponseEntity<SettingsResponse> settings() {
     }
 
     @GetMapping("/tag")
-    private ResponseEntity<TagResponse> tag(@RequestParam(required = false, defaultValue = "") String query) {
+    private ResponseEntity<TagResponse> tag(@RequestParam(required = false) Optional<String> query) {
```
Comment (Klepus) [inline src/main/java/com/javlasov/blog/controller/ApiGeneralController.java:41]: почему private? как тестировать потом будешь?

File: src/main/java/com/javlasov/blog/dto/PostCommentDto.java [hunk 4]
```diff
@@ -10,6 +10,6 @@ public class PostCommentDto {
     private int id;
     private long timestamp;
     private String text;
-    private UserDto user;
+    private UserDtoForPosts user;
```
Comment (Klepus) [inline src/main/java/com/javlasov/blog/dto/PostCommentDto.java:13]: мб UserPostsDto?

**Review by aleks-klepinin — CHANGES_REQUESTED** [review]

**Pull request merged**